google/cloud/spanner_dbapi/connection.py

```diff
@@ -190,6 +265,33 @@ def run_prior_DDL_statements(self):
 
             return self.database.update_ddl(ddl_statements).result()
 
+    def run_statement(self, statement, retried=False):
+        """Run single SQL statement in begun transaction.
+
+        This method is never used in autocommit mode. In
+        !autocommit mode however it remembers every executed
+        SQL statement with its parameters.
+
+        :type statement: :class:`dict`
+        :param statement: SQL statement to execute.
+
+        :rtype: :class:`google.cloud.spanner_v1.streamed.StreamedResultSet`,
+                :class:`google.cloud.spanner_dbapi.checksum.ResultsChecksum`
+        :returns: Streamed result set of the statement and a
+                  checksum of this statement results.
+        """
+        transaction = self.transaction_checkout()
+        self._statements.append(statement)
```

Comment:
I think we should only add the statement here if `retried=False`. Otherwise, if the transaction is aborted and retried multiple times, the list of statements that are executed on the transaction is duplicated.

Reply:
Oh, yeah, I'm only clearing statements on `commit()` and `rollback()` for now. Good catch!